core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala

```diff
@@ -344,12 +354,20 @@ private[spark] class MesosSchedulerBackend(
 
   override def frameworkMessage(d: SchedulerDriver, e: ExecutorID, s: SlaveID, b: Array[Byte]) {}
 
+  /**
+   * Remove executor associated with slaveId in a thread safe manner.
+   */
+  private def removeExecutor(slaveId: String) = {
+    synchronized {
```

Comment:
Yes, it is non-consistent. The synchronized block was just carried over from the previous code. We could probably use SynchronizedSet to prevent concurrentmodification exceptions. There is a synchronized {} in statusUpdate to try to ensure atomicity modifying slaveIdsWithExecutors and taskIdToSlaveId but that isn't enough since slaveIdsWithExecutors is modified in a number of places.